### PR TITLE
Implemented Feature for encrypting Email before storing in DB.

### PR DESCRIPTION
**Pull request by Anubhav-2003**
**What kind of change does this PR introduce?**
A Feature.

**Issue Number:**
Fixes : #1756 

**Did you add tests for your changes?**
Yes.

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**
In the following Pull-Request, I have implemented the security feature of encrypting the Emails, before storing them in our Database. This PR would enhance the security, and privacy of our users substantially.  I have updated all the test-cases, and wrote new ones where necessary.

**Does this PR introduce a breaking change?**
Yes. All current contributors, must re-import the sample data after merging this PR for smooth functionality.

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

**Comment by Cioppolo14** [comment]
@Anubhav-2003 Please fix the failed tests.

**Comment by Anubhav-2003** [comment]
@palisadoes 
Respected sir, 

I have completely implemented the algorithm for encrypting the emails before storing in the Database. There was a PR revert in the develop branch while I was working on this feature-branch. Therefore, I had to merge the updated develop with my feature branch. 

My PR is failing the linting tests because a lot of those files involved in the PR revert had linting errors already present, as a consequence hundreds of lines of lint errors are being shown to me. Majority of which I have not written code for. 

A separate issue must be made to tackle all the lint errors in the code base, and if there is one, it must be accelerated otherwise hundreds of lint errors come for every merge from the develop depending upon the number of files changed.

Thank You.

**Comment by Anubhav-2003** [comment]
@Cioppolo14 Respected ma'am, 

Actually I have written the reason for the failing tests above, I will still try to fix as many lint errors as possible but most of them are for code I did not write, and there are hundreds of pre existing errors in the codebase.

Thank You.

**Comment by palisadoes** [comment]
1. I have to close this PR. 
2. You have removed a key portion of our app, the importation of test data for people to use.
3. This means that future aspects of testing will fail.
4. Please fix the importation of the setup data to match the expected functionality.

**Comment by Anubhav-2003** [comment]
@palisadoes Respected sir,

Last I checked the Importation of sample data was working fine. If you are referring to the function I commented out in the setup. That was because the function was never being called. While I was working on the issue, a PR was reverted as a consequence there were large no. of merge conflicts for parts of the file, I did not write code for. I just kept the incoming changes from develop.

Could you please specify which part you are referring too. As I ran tests locally and most of them passed.

Thank you.

**Comment by palisadoes** [comment]
image

**Comment by Anubhav-2003** [comment]
@palisadoes Respected sir, I was talking about this function only. I commented this function out because this function was never actually invoked and was never actually used but was just written. I think it had something to do with the previous PR but after the revert this function became unused so I commented it for future references. Thank you.

**Comment by palisadoes** [comment]
Which PR are you talking about?

**Comment by Anubhav-2003** [comment]
@palisadoes Sir this one https://github.com/Anubhav-2003/talawa-api/commit/6295a238f26643542119c6c3e30cd841e147433c#diff-bfe5fcdcbcc5b806d61f26909a1ef6cc3b98d63016e6243c3f0326ec4d28be73

In fact if you look closely sir, the function, I commented is not even part of our current develop branch.  That is why I commented it. I think there was some bug with this functionality as a result it was causing user registration problems in the admin portal so I guess you reverted it.

Thank you.

**Comment by palisadoes** [comment]
OK. Reopening. Thanks for the information.

**Comment by Anubhav-2003** [comment]
> OK. Reopening. Thanks for the information.

Thank you a lot sir. For your understanding.

**Comment by Anubhav-2003** [comment]
@palisadoes Respected sir, 

The PR looks good to merge. Please review. The test cases are getting failed due to the fact that no encryption key is set in the cloud instance running the tests, This is because the workflow file is directly running 'npm test' without running npm setup. And the encryption key is set during the setup. Without the key tests cannot pass. Locally all tests are getting passed.

I would suggest that after this PR is merged, to create a new issue to update the pull-request.yml file to run the setup as well to initialize the encryption key before the tests can take place.

I would be more than happy to solve the issue as well. Or we can update the file in this issue only without creating a new one. As you would suggest.

Thank You.

**Comment by palisadoes** [comment]
1. This needs to be fixed in this PR. 
1. All tests must pass as a criteria for merging. The tests must be valid
1. Please make the appropriate changes.
1. We cannot assign reviewers otherwise

**Comment by Anubhav-2003** [comment]
> 1. This needs to be fixed in this PR.
> 2. All tests must pass as a criteria for merging. The tests must be valid
> 3. Please make the appropriate changes.
> 4. We cannot assign reviewers otherwise

Ok sir.

**Comment by Anubhav-2003** [comment]
@palisadoes  Respected sir,

I have completely implemented the feature. Please review.

-------------------------------------------------------------------------------------------------------------------------------------

I also wanted to give an insight into the process I am using for encrypting the emails:

1.  First during the initial setup of the API, an encryption key is set as an environmental variable (a single master key for all emails).

2. Everytime, an email is encrypted, a random salt is used for encryption. rather than storing the salt separately as field in the DB, as many orgs do, I have prepended the salt to the encrypted email to add an extra layer of security. As a consequence even if someone gets access to our DB and the security key. They will not be able to decrypt the email unless they know the SALT_LENGTH.

3. The salt length currently has been set to 16, but when going into production it must be stored using a KMS like vault from HashICorp.

Thank You.

**Comment by palisadoes** [comment]
1. This isn't ready for review. The tests on your patch aren't sufficiently covered.
2. Let us know when it's ready

File: .eslintrc.json [hunk 1]
```diff
@@ -34,8 +34,8 @@
       // which has convulated return types
       "files": ["./src/index.ts", "./src/utilities/copyToClipboard.ts"],
       "rules": {
-        "@typescript-eslint/explicit-function-return-type": "off",
-        "@typescript-eslint/no-empty-function": "off"
+        "@typescript-eslint/explicit-function-return-type": "error",
+        "@typescript-eslint/no-empty-function": "error"
       }
```
Comment (SiddheshKukade) [inline .eslintrc.json:39]: Why have you turned on  linter for this ?

File: README.md [hunk 2]
```diff
@@ -25,11 +25,10 @@ Core features include:
...
 
-- [Talawa API](#talawa-api)
-  - [Talawa Components](#talawa-components)
-  - [Documentation](#documentation)
-  - [Installation](#installation)
-  - [Image Upload](#image-upload)
+- [Talawa Components](#talawa-components)
+- [Documentation](#documentation)
+- [Installation](#installation)
+- [Image Upload](#image-upload)
```
Comment (SiddheshKukade) [inline README.md:31]: No need to change this file

File: .eslintrc.json (hunk 1 above)
Comment (Anubhav-2003) [reply .eslintrc.json:39]: Respected sir,  Actually I was checking certain lint errors. Let me revert those changes. Thank you.

File: README.md (hunk 2 above)
Comment (Anubhav-2003) [reply README.md:31]: Actually sir, I did not change this file. Maybe it was due to a previous PR that got reverted while I was working on this feature branch. Thank You.

**Comment by Anubhav-2003** [comment]
> 1. This isn't ready for review. The tests on your patch aren't sufficiently covered.
> 
>     2. Let us know when it's ready

Ok sir. Let me increase the coverage. Thank you.

**Comment by Anubhav-2003** [comment]
@palisadoes Respected sir,

The PR is ready to merge. Please review. 

I have maximized the test coverage as much as possible, and it is above thresh hold, the only few lines missing coverage are actually the non-null assertion 'if statements' checks which typescript enforces. Tests for them are not needed.  I have written or updated new tests for every new line of code written.

Thank You.

**Comment by palisadoes** [comment]
1. In setup.ts the max image size is prompted for after the database importation. 
2. It should be prompted earlier in the process. Preferably as the second prompt after the environment step.

**Comment by Anubhav-2003** [comment]
@palisadoes Respected Sir,

I will make the changes as requested by you, but I wanted to point out is that the sequence is currently part of our develop branch and was not influenced by this pull request. Therefore it would be more appropriate if I make that change in the issue #1680 related to setup.ts that I have been assigned. As this thing is not part of the current issue related to this pull request.

I will make that change as part of the PR for #1680

Thank you.

**Comment by palisadoes** [comment]
> @palisadoes Respected Sir,
> 
> I will make the changes as requested by you, but I wanted to point out is that the sequence is currently part of our develop branch and was not influenced by this pull request. Therefore it would be more appropriate if I make that change in the issue #1680 related to setup.ts that I have been assigned. As this thing is not part of the current issue related to this pull request.
> 
> I will make that change as part of the PR for #1680
> 
> Thank you.

OK, fix it in the other issue.

**Comment by Anubhav-2003** [comment]
@palisadoes Also sir, I would like to request , the sooner this PR is reviewed and merged the more optimised it will be, as every day. The PR feature branch is getting outdated as compared to the current develop. which might lead to unnecessary merge conflicts.

Thank you.

**Comment by palisadoes** [comment]
Please merge with the latest upstream

**Comment by palisadoes** [comment]
@SiddheshKukade @SiddheshKukade Please review

**Comment by Anubhav-2003** [comment]
@palisadoes ok sir. Thank you.

**Comment by palisadoes** [comment]
1. You have disabled linting in the eslint config file and other locations. 
1. Please restore

You will degrade the quality of our code base with this approach

**Comment by Anubhav-2003** [comment]
@palisadoes Sorry sir for the mistake, I don’t know how it happened. Let me restore it.

 But one thing that I wanted to point out is that there are literally hundreds of linting errors, pre-existing in our code base mostly concerned with the any type. Now, unless those errors are fixed. The quality of code in the code base could not be improved, because if a new developer contributes to a pre-existing file, he encounters hundreds of errors as a result of which he is not able to commit to git. 

So as a result, the only thing he resorts to is disabling lint using comments. Because there is simply no other way to commit. So, unless every last lint error in the code base is fixed, enforcing such strict types is just resulting in more comments. So my suggestion would be sir to create a top priority issue to 1st solve every last lint error. 

One more thing sir that I found as a fact for extensive use of the any type by developers is that there is no single point of truth for the objects and the interface types. Since we are using a graphQL based API we should be using object inferring libraries like Zod,  which would create a single point of truth, for both the front-end and the backend so that just a change at one place would ensure the structure is changed everywhere, and would reduce the use of any type. Which could be incorporated even more efficiently using a monolith approach rather than discrete repositories. In the last few issues, I have worked. I realised that just changing one parameter in the organisation. Object resulted in me changing dozens of files for repeated and recurring stuff. This actually forces developers to just use the any type.

I am extremely sorry sir. If I made any mistake with the linting, I will restore it right away. But some very drastic changes are needed to be made the code base if we really want to make the Code error free.

Thank You.

**Comment by palisadoes** [comment]
Create issues to use `zod` with an explanation

**Comment by Anubhav-2003** [comment]
@palisadoes Ok sir, thank you. And sorry again.

**Comment by palisadoes** [comment]
> @palisadoes Respected Sir,
> 
> I have changed the option in the ESLint file from “off” to error as per your order. But apart from it every Eslint comment that I have added are actually for existing lint errors of those files. Without adding those, it would be virtually impossible for me to commit to git, because then I would have to change the options in the configuration files.
> 
> Every new line of code that I have written, for instance in encryptionmodule.ts and others have been passed through lint checks, as well prettier. If you still feel this would be degrading the code quality, then please suggest what should I do.
> 
> Also, as per your suggestion and guidance, I am working on a proposal for implementation of ZOD. But implementing it would not be trivial as it would involve changes in multiple repositories, along with change in the system design, and would at the least take 2 months.. So I am planning for an elaborate proposal.
> 
> Thank you.

I spoke with one of the reviewers. They feel that `zod` isn't required in the immediate future and they too feel it will take a lot of time. Please discuss this with them. 

I'd prefer to focus on adding user facing features for a working prototype. That is a much greater priority.

**Comment by xoldyckk** [comment]
Zod is extremely slow, if it is to be implemented I'd suggest [typia](https://github.com/samchon/typia) or [typebox](https://github.com/sinclairzx81/typebox).

**Comment by Anubhav-2003** [comment]
> > @palisadoes Respected Sir,
> > I have changed the option in the ESLint file from “off” to error as per your order. But apart from it every Eslint comment that I have added are actually for existing lint errors of those files. Without adding those, it would be virtually impossible for me to commit to git, because then I would have to change the options in the configuration files.
> > Every new line of code that I have written, for instance in encryptionmodule.ts and others have been passed through lint checks, as well prettier. If you still feel this would be degrading the code quality, then please suggest what should I do.
> > Also, as per your suggestion and guidance, I am working on a proposal for implementation of ZOD. But implementing it would not be trivial as it would involve changes in multiple repositories, along with change in the system design, and would at the least take 2 months.. So I am planning for an elaborate proposal.
> > Thank you.
> 
> I spoke with one of the reviewers. They feel that `zod` isn't required in the immediate future and they too feel it will take a lot of time. Please discuss this with them.
> 
> I'd prefer to focus on adding user facing features for a working prototype in the next two months vs spending it on `zod`. That is a much greater priority.

Ok sir, we will work on those features according to your request. As features that the users actually interact with are definitely more important. Is there anything that I have to do in this PR?

Thank you.

**Comment by palisadoes** [comment]
@Anubhav-2003 If you feel this is a valid need and you are interested in GSoC, please discuss this with the mentor team if we get selected.

**Comment by Anubhav-2003** [comment]
> Zod is extremely slow, if it is to be implemented I'd suggest [typia](https://github.com/samchon/typia) or [typebox](https://github.com/sinclairzx81/typebox).

Ok sir, if in future, we decide to do object validation and inferences, then I would implement it using typia or typebox as per your guidance. Also sir, my review for this PR is pending for a few days. Could you please give a feedback.

Thank you.

**Comment by Anubhav-2003** [comment]
> @Anubhav-2003 If you feel this is a valid need and you are interested in GSoC, please discuss this with the mentor team if we get selected.

Ok sir. Actually, I’m already working on a proposal, which I think would benefit the API more than this object object inference by Zod. I had actually messaged you on Slack regarding this, and you told me to talk with the mentors once we get selected. So I am currently working on the idea and the proposal so that I can interact with the mentors once we get finalised. Thank you.

**Comment by Anubhav-2003** [comment]
> 1. Consider the feedback from @xoldyckk on whether `zod` is appropriate
> 2. I will not provide any further feedback on the matter. I have to stay neutral, the mentors make the final decisions.

Ok sir. Actually My idea is not related to zod, but related to how static files are handled by our API after uploading, as they are just stored in the code tree with a .gitignore. So I am working on a proposal for improvising the method to production grade. I will be contacting @xoldyckk and other mentors once my proposal is ready. 

Thank you sir for your guidance.

**Comment by xoldyckk** [comment]
First of all emails aren't meant to be encrypted, this would only create plethora of other problems. Email is a piece of data for uniquely identifying a particular user. 

Besides, when we're using a cloud database like mongodb atlas, we should trust their security more than our own. Centralizing the encryption key for all the users of the talawa-api is the same amount of security risk in my honest opinion. Once it is leaked the hacker once again

**Comment by SiddheshKukade** [comment]
@Anubhav-2003 Please take pull from origin develop so that I can review

**Comment by Anubhav-2003** [comment]
@SiddheshKukade Respected sir, 

I have updated the feature branch with the latest upstream as per your request. But the thing is that upstream develop is being updated every few minutes so chances are it will again be out-of-date by the time you see the message. Also sir all the tests were getting passed initially but after merging the latest develop, 1 test file has failed. 

Bu the test that is failing in the github actions, is actually passing on my device locally. It would be helpful if you could suggest what to do. As the test seems to have been affected by the latest changes in develop. I am attaching a screenshot of the test passing locally as well.  

Thank You.
Screenshot from 2024-02-18 12-44-08

**Comment by Anubhav-2003** [comment]
@SiddheshKukade Respected sir,

I have updated the feature branch with the latest upstream as per your request. The thing is sir, after merging the latest upstream develop, one test seems to be failing as initially all the tests were passing. It seems that the latest develop branch changes have affected it in some way. But the test that is failing in github actions is actually passing on my local machine. Let me see how and why it is happening, or it would be also highly valuable for me if you could suggest something.

I am also  attaching a screenshot of the test, that it is passing locally.

Thank You.
Screenshot from 2024-02-18 12-44-08

**Comment by Anubhav-2003** [comment]
@EshaanAgg Respected sir,

As you can see the only test case failing now is the forgotpassword.spec.tsx, a file I have not made any changes into and is failing after a recent pull request was reverted in the develop branch and has nothing to do with this PR. This is the PR https://github.com/PalisadoesFoundation/talawa-api/commit/a083377ecf71f26d3463ed0d954f973e8c439241 I am talking about. Before this all the tests were being passed.

Can you please review my Pull request as it has been pending for quite some time now. 

Thank you.

**Comment by Anubhav-2003** [comment]
Respected sir, 

I have fixed all the issue. Including one of the issues that happened due to this PR revert a083377. Please review.

Thank You.

**Comment by Cioppolo14** [comment]
@Anubhav-2003 Can you fix the code coverage tests?

**Comment by Cioppolo14** [comment]
@SiddheshKukade Can you look at this PR?

**Comment by Anubhav-2003** [comment]
> @Anubhav-2003 Can you fix the code coverage tests?

Respected ma'am, 

I have already maximized the code coverage possible. The only lines of code missing coverage are actually extra non-null type assertions, that typescript forces. But are actually covered practically. 

Thank You.

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by Cioppolo14** [comment]
@Anubhav-2003 Please fix the failing tests.

**Comment by palisadoes** [comment]
Please fix the conflicting files, merge with the upstream too

**Comment by palisadoes** [comment]
This is an update on the PR merging freeze:

1. In the next few hours we will be merging the `develop-userTypeFix` branch into the `develop` branch. 
1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. This merge will cause some conflicts in your PR.

We decided to do this at the beginning of the weekend to give us all time to adjust PR code and create bug fixes that may arise.

Update your code at or after midnight GMT on the morning of March 23, 2024. (5:30am IST). 

If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by palisadoes** [comment]
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by Cioppolo14** [comment]
@Anubhav-2003 Please fix the conflicting files and failed tests.

**Comment by Anubhav-2003** [comment]
@palisadoes Respected sir, After the latest PR merge of usertype-fix branch, the introduction of AppUserProfile and the changes in the structure of the User field, the implementation that I had adopted in this PR is practically rendered null as the merge conflicts are too huge, and the overall implementation of this PR as well will not work. I tried to update this PR, but don't think this present one could now be merged. Should I create a new one for this issue based on the current develop? As a lot of things have changed since this PR was raised. Thank You.

**Comment by palisadoes** [comment]
Yes

**Comment by palisadoes** [comment]
Closing as abandoned.

**Pull request closed without merging**